File: src/import/csv.ts
/**
 * Low-level CSV parsing: delimiter sniffing and raw-matrix extraction via `papaparse`.
 * No column semantics here — see `columns.ts` for structural role detection.
 *
 * Row-splitting itself is cheap even at several thousand rows (milliseconds); the parts
 * of the import pipeline that could plausibly stall the main thread — per-row hashing and
 * categorisation — are chunked with explicit yields in `parse.ts` instead of here.
 */
import Papa from 'papaparse';

export interface RawCsv {
  /** Every row as an array of trimmed string cells. Ragged rows are padded with ''. */
  rows: string[][];
  delimiter: string;
  /** Parse errors papaparse surfaced (bad quoting, etc.) — worth showing as warnings. */
  errors: string[];
}

/** Parse raw CSV/TSV text into a rectangular matrix of trimmed string cells. */
export function parseRawCsv(text: string): RawCsv {
  const result = Papa.parse<string[]>(text, {
    header: false,
    skipEmptyLines: 'greedy',
    delimitersToGuess: [',', ';', '\t', '|'],
  });

  const width = result.data.reduce((max, row) => Math.max(max, row.length), 0);
  const rows = result.data
    .filter((row) => row.length > 0)
    .map((row) => {
      const padded = row.slice(0, width).map((cell) => (cell ?? '').trim());
      while (padded.length < width) padded.push('');
      return padded;
    });

  return {
    rows,
    delimiter: result.meta.delimiter || ',',
    errors: result.errors.map((e) => `Row ${e.row ?? '?'}: ${e.message}`),
  };
}

/**
 * Read a `File` (from a drop/pick input) as text. Wrapped so the import feature never
 * touches `FileReader` directly.
 */
export async function readFileAsText(file: File): Promise<string> {
  return file.text();
}

/** Yield control back to the event loop / paint. Used to keep large-file processing from freezing the UI. */
export function yieldToUi(): Promise<void> {
  return new Promise((resolve) => {
    const w = typeof window !== 'undefined' ? window : undefined;
    if (w && 'requestAnimationFrame' in w) {
      w.requestAnimationFrame(() => resolve());
    } else {
      setTimeout(resolve, 0);
    }
  });
}

File: android/app/src/main/java/com/tally/app/csvimport/Csv.kt
package com.tally.app.csvimport

/**
 * Low-level CSV parsing: delimiter sniffing and raw-matrix extraction.
 * Ported from src/import/csv.ts (which used papaparse; this is a from-scratch
 * Kotlin tokenizer since no CSV library dependency is allowed in this layer).
 * No column semantics here — see Columns.kt for structural role detection.
 */
data class RawCsv(
    /** Every row as a list of trimmed string cells. Ragged rows are padded with "". */
    val rows: List<List<String>>,
    val delimiter: Char,
    /** Parse errors worth showing as warnings. Always empty for this tokenizer
     *  (it never throws on malformed quoting — ragged/ambiguous input degrades
     *  gracefully into extra/short fields instead). */
    val errors: List<String>
)

private val CANDIDATE_DELIMITERS = listOf(',', ';', '\t', '|')

/** Count [delimiter] occurrences in [line] that are outside a quoted span. */
private fun countUnquoted(line: String, delimiter: Char): Int {
    var count = 0
    var inQuotes = false
    for (c in line) {
        if (c == '"') {
            inQuotes = !inQuotes
        } else if (c == delimiter && !inQuotes) {
            count++
        }
    }
    return count
}

/**
 * Guess the delimiter from a sample of non-blank lines: for each candidate,
 * find the modal (most common) per-line count, and score it by how
 * consistently that count recurs, weighted by the count itself (so a
 * delimiter that appears 4 times per line consistently beats one that
 * appears once per line consistently — a wider matrix is a stronger signal).
 */
private fun sniffDelimiter(sampleLines: List<String>): Char {
    var best = ','
    var bestScore = -1.0
    for (d in CANDIDATE_DELIMITERS) {
        val counts = sampleLines.map { countUnquoted(it, d) }
        if (counts.isEmpty() || counts.all { it == 0 }) continue
        val mode = counts.groupingBy { it }.eachCount().maxByOrNull { it.value } ?: continue
        if (mode.key == 0) continue
        val consistency = mode.value.toDouble() / counts.size
        val score = consistency * mode.key
        if (score > bestScore) {
            bestScore = score
            best = d
        }
    }
    return best
}

/** RFC4180-ish tokenizer: [delimiter]-separated fields, `"`-quoted with `""` escaping,
 *  quoted fields may contain the delimiter or newlines. `\n` outside quotes ends a row. */
private fun tokenizeCsv(text: String, delimiter: Char): List<List<String>> {
    val rows = mutableListOf<List<String>>()
    var row = mutableListOf<String>()
    val field = StringBuilder()
    var inQuotes = false
    var i = 0
    val n = text.length

    while (i < n) {
        val c = text[i]
        if (inQuotes) {
            when {
                c == '"' && i + 1 < n && text[i + 1] == '"' -> {
                    field.append('"')
                    i += 2
                }
                c == '"' -> {
                    inQuotes = false
                    i++
                }
                else -> {
                    field.append(c)
                    i++
                }
            }
        } else {
            when (c) {
                '"' -> {
                    inQuotes = true
                    i++
                }
                delimiter -> {
                    row.add(field.toString())
                    field.clear()
                    i++
                }
                '\n' -> {
                    row.add(field.toString())
                    field.clear()
                    rows.add(row)
                    row = mutableListOf()
                    i++
                }
                else -> {
                    field.append(c)
                    i++
                }
            }
        }
    }
    if (field.isNotEmpty() || row.isNotEmpty()) {
        row.add(field.toString())
        rows.add(row)
    }
    return rows
}

/**
 * Parse raw CSV/TSV text into a rectangular matrix of trimmed string cells.
 * Delimiter is sniffed from `[',', ';', '\t', '|']`. Blank/whitespace-only
 * rows are dropped ("greedy" empty-line skipping, matching papaparse's
 * `skipEmptyLines: 'greedy'`).
 */
fun parseRawCsv(text: String): RawCsv {
    val normalized = text.replace("\r\n", "\n").replace("\r", "\n")
    val allLines = normalized.split("\n")
    val sampleLines = allLines.filter { it.isNotBlank() }.take(10)
    val delimiter = sniffDelimiter(sampleLines)

    val rawRows = tokenizeCsv(normalized, delimiter)
    val nonEmptyRows = rawRows.filter { row -> row.any { it.trim().isNotEmpty() } }

    val width = nonEmptyRows.maxOfOrNull { it.size } ?: 0
    val rows = nonEmptyRows.map { row ->
        val cells = row.take(width).map { it.trim() }.toMutableList()
        while (cells.size < width) cells.add("")
        cells.toList()
    }

    return RawCsv(rows, delimiter, emptyList())
}
